### DISCUSSION TITLE
How to show several images with different aspect ratio?

### DISCUSSION THREAD

**Discussion by Frank1126lin** [slint-ui/slint#2996] [Q&A] [answered]
I am writing the GUI software on WIN 11 with Rust, thus I have different kinds of images to show, with different aspect ratio, how I lock the aspect ratio to show the images. 
If I write as below, the images apect ratio was changed.
```
Image {
            width:  100%;
            height: 100%;
            source: root.img;
        }
```
Well, if I just write the width, some of the images shows right, but with a long image, the height is much bigger than the rectangle.
```
 Image {
            width:  100%;
            source: root.img;
            // implicit default, preserving aspect ratio:
            // height: self.width * natural_height / natural_width;
        }
```
How do I fix that? anyone can help?

**Comment by hunger** [comment]
Here is a link to Slintpad with a [demo code snippet](https://slint.dev/releases/1.1.0/editor/?snippet=import+%7B+Button%2C+VerticalBox+%7D+from+%22std-widgets.slint%22%3B%0Aexport+component+Demo+%7B%0A++++height%3A+500px%3B%0A++++width%3A+500px%3B%0A++++Image+%7B%0A++++++++source%3A+%40image-url%28%22https%3A%2F%2Fslint.dev%2Flogo%2Fslint-logo-full-light.svg%22%29%3B%0A++++++++height%3A+100%25%3B%0A++++++++width%3A+100%25%3B%0A++++++++image-fit%3A+contain%3B%0A++++%7D%0A%7D%0A).

I used the SLint logo, which is wider than it is high, but the same applies in the other direction, too.

If you remove the `width: 100%;`, then the image is free to select whatever width is appropriate for the given height (with parts of it being cut away again).

When you constrain both width and height, then the image will be trimmed down to fit into the assign space. Note that I had to specify `image-fit` to tells lint to preserve the aspect ratio.

**Reply by Frank1126lin** [reply]
Well,what I want to do is to show the different images in it's own aspect ratio, How can I lock the aspect ratio in slint?
For Example, I write a sample in slintpad like [this](https://slint.dev/releases/1.0.0/editor/?snippet=import+%7B+Button%2C+VerticalBox+%7D+from+%22std-widgets.slint%22%3B%0A%0Acomponent+Photo+%7B%0A++++in+property+%3Cimage%3E++img%3B%0A++++preferred-height%3A+200px%3B%0A++++preferred-width%3A+300px%3B%0A++++padding%3A+10px%3B%0A++++Rectangle+%7B%0A++++++++border-width%3A+1px%3B%0A++++++++border-color%3A+black%3B%0A++++++++border-radius%3A+1px%3B%0A++++++++Image+%7B%0A++++++++++++width%3A++100%25%3B%0A++++++++++++height%3A+100%25%3B%0A++++++++++++source%3A+img%3B%0A++++++++%7D%0A++++++++%7D%0A%7D%0Aexport+component+App+inherits+Window+%7B%0A+GridLayout+%7B%0A++++++++++++spacing%3A+10px%3B%0A++++++++++++Row+%7B%0A++++++++++++++++Photo%7B%0A++++++++++++++++++++img%3A+%40image-url%28%22https%3A%2F%2Fslint.dev%2Flogo%2Fslint-logo-full-light.svg%22%29%3B%0A++++++++++++++++%7D%0A%0A++++++++++++++++Photo%7B%0A++++++++++++++++++++img%3A+%40image-url%28%22https%3A%2F%2Fslint.dev%2Fassets%2Fimg%2Fslint-favicon.svg%22%29%3B%0A++++++++++++++++%7D%0A++++++++++++%7D%0A%7D%0A%7D%0A).
I want to show both of them in one type component, with its own aspect ratio, but I cann't do it in slint. The image's aspect ratio was changed if I write both width and height. otherwise, if I only write the width, the second image was too high to show in the box.

**Reply by ogoffart** [reply]
Do you mean something like
```
height: self.width * self.source.height / self.source.width;
```

**Reply by Frank1126lin** [reply]
Yeah, I tried, it works in keep the aspect ratio, but the height still longer than the box in this [slintpad demo](https://slint.dev/releases/1.0.0/editor/?snippet=import+%7B+Button%2C+VerticalBox+%7D+from+%22std-widgets.slint%22%3B%0A%0Acomponent+Photo+%7B%0A++++in+property+%3Cimage%3E++img%3B%0A++++preferred-height%3A+200px%3B%0A++++preferred-width%3A+300px%3B%0A++++padding%3A+10px%3B%0A++++Rectangle+%7B%0A++++++++border-width%3A+1px%3B%0A++++++++border-color%3A+black%3B%0A++++++++border-radius%3A+1px%3B%0A++++++++Image+%7B%0A++++++++++++width%3A++100%25%3B%0A++++++++++++height%3A+self.width+*+self.source.height+%2F+self.source.width%3B%0A++++++++++++source%3A+img%3B%0A++++++++%7D%0A++++++++%7D%0A%7D%0Aexport+component+App+inherits+Window+%7B%0A+GridLayout+%7B%0A++++++++++++spacing%3A+10px%3B%0A++++++++++++Row+%7B%0A++++++++++++++++Photo%7B%0A++++++++++++++++++++img%3A+%40image-url%28%22https%3A%2F%2Fslint.dev%2Flogo%2Fslint-logo-full-light.svg%22%29%3B%0A++++++++++++++++%7D%0A%0A++++++++++++++++Photo%7B%0A++++++++++++++++++++img%3A+%40image-url%28%22https%3A%2F%2Fslint.dev%2Fassets%2Fimg%2Fslint-favicon.svg%22%29%3B%0A++++++++++++++++%7D%0A++++++++++++%7D%0A%7D%0A%7D%0A).
So, is there anyway to set the max width or max height for these images? thus these images can be set by its own to keep the size and aspect ratio.

**Reply by hunger** [reply, marked as answer]
@Frank1126lin: Add `image-fit: contain;` into the `Image` in your `Photo`.

[Slintpad](https://slint.dev/releases/1.0.0/editor/?snippet=import+%7B+Button%2C+VerticalBox+%7D+from+%22std-widgets.slint%22%3B%0A%0Acomponent+Photo+%7B%0A++++in+property+%3Cimage%3E++img%3B%0A++++preferred-height%3A+200px%3B%0A++++preferred-width%3A+300px%3B%0A++++padding%3A+10px%3B%0A++++Rectangle+%7B%0A++++++++border-width%3A+1px%3B%0A++++++++border-color%3A+black%3B%0A++++++++border-radius%3A+1px%3B%0A++++++++Image+%7B%0A++++++++++++width%3A++100%25%3B%0A++++++++++++height%3A+100%25%3B%0A++++++++++++source%3A+img%3B%0A++++++++++++image-fit%3A+contain%3B%0A++++++++%7D%0A++++%7D%0A%7D%0Aexport+component+App+inherits+Window+%7B%0A+GridLayout+%7B%0A++++++++++++spacing%3A+10px%3B%0A++++++++++++Row+%7B%0A++++++++++++++++Photo%7B%0A++++++++++++++++++++img%3A+%40image-url%28%22https%3A%2F%2Fslint.dev%2Flogo%2Fslint-logo-full-light.svg%22%29%3B%0A++++++++++++++++%7D%0A%0A++++++++++++++++Photo%7B%0A++++++++++++++++++++img%3A+%40image-url%28%22https%3A%2F%2Fslint.dev%2Fassets%2Fimg%2Fslint-favicon.svg%22%29%3B%0A++++++++++++++++%7D%0A++++++++++++%7D%0A%7D%0A%7D%0A)

**Reply by bjorn** [reply]
I'm unable to reproduce @Frank1126lin's issue, even when [explicitly specifying `ImageFit.fill`](https://slint.dev/releases/1.1.0/editor/?snippet=import+%7B+Button%2C+VerticalBox+%7D+from+%22std-widgets.slint%22%3B%0A%0Acomponent+Photo+%7B%0A++++in+property+%3Cimage%3E++img%3B%0A++++preferred-height%3A+200px%3B%0A++++preferred-width%3A+300px%3B%0A++++padding%3A+10px%3B%0A++++Rectangle+%7B%0A++++++++border-width%3A+1px%3B%0A++++++++border-color%3A+black%3B%0A++++++++border-radius%3A+1px%3B%0A%0A++++++++Image+%7B%0A++++++++++++width%3A++100%25%3B%0A++++++++++++height%3A+100%25%3B%0A++++++++++++source%3A+img%3B%0A++++++++++++image-fit%3A+fill%3B%0A++++++++%7D%0A++++%7D%0A%7D%0Aexport+component+App+inherits+Window+%7B%0A+GridLayout+%7B%0A++++++++++++spacing%3A+10px%3B%0A++++++++++++Row+%7B%0A++++++++++++++++Photo%7B%0A++++++++++++++++++++img%3A+%40image-url%28%22https%3A%2F%2Fslint.dev%2Flogo%2Fslint-logo-full-light.svg%22%29%3B%0A++++++++++++++++%7D%0A%0A++++++++++++++++Photo%7B%0A++++++++++++++++++++img%3A+%40image-url%28%22https%3A%2F%2Fslint.dev%2Fassets%2Fimg%2Fslint-favicon.svg%22%29%3B%0A++++++++++++++++%7D%0A++++++++++++%7D%0A%7D%0A%7D%0A). What could be causing that?

![image](https://github.com/slint-ui/slint/assets/531764/df05929a-0ae0-42ce-b018-421fd7f63fd3)

I was a little confused by the problem and the provided solution, since to me the result looks exactly the same.

**Reply by hunger** [reply]
This bit of code:

```
        Image {
            width:  100%;
            height: 100%;
            source: img;
            image-fit: contain;
        }
```

renders this:

![image](https://github.com/slint-ui/slint/assets/73267/b38ce627-41e4-4358-9f2c-6ada34fe38b3)

This is the code I got from @Frank1126lin:

```
        Image {
            width:  100%;
            height: self.width * self.source.height / self.source.width;
            source: img;
        }
```

which renders like this:

![image](https://github.com/slint-ui/slint/assets/73267/13abd1aa-c9cf-4de1-b319-c0d3b1f73b5d)

**Reply by Frank1126lin** [reply]
Yep, I got the right result using image-fit: contain, Thanks for help. by the way, what I wonder was that is the contain was default set by Doc?
image-fit (in enum [ImageFit](https://slint.dev/releases/1.0.0/docs/slint/src/builtins/enums.html#imagefit)): Specifies how the source image shall be fit into the image element. (default value: contain when the Image element is part of a layout, fill otherwise)

**Reply by bjorn** [reply]
@hunger I get that, but what I was saying, is that I can't reproduce this reported behavior: "If I write as below, the images apect ratio was changed.". It seems on my end (testing [this on Firefox](https://slint.dev/releases/1.1.0/editor/?snippet=import+%7B+Button%2C+VerticalBox+%7D+from+%22std-widgets.slint%22%3B%0Aexport+component+Demo+%7B%0A++++height%3A+500px%3B%0A++++width%3A+500px%3B%0A++++Image+%7B%0A++++++++source%3A+%40image-url%28%22https%3A%2F%2Fslint.dev%2Flogo%2Fslint-logo-full-light.svg%22%29%3B%0A++++++++height%3A+100%25%3B%0A++++++++width%3A+100%25%3B%0A++++++++image-fit%3A+fill%3B%0A++++%7D%0A%7D%0A)), I can't get it to show images with changed aspect ratio at all.

**Reply by hunger** [reply]
@bjorn: Strange...

This code:

```
        Image {
            width:  100%;
            height: 100%;
            source: img;
            image-fit: fill;
        }
```

gets me this:

![image](https://github.com/slint-ui/slint/assets/73267/dcd428f9-198f-46ae-b9b5-df885ee2fe3f)

Firefox or not should not matter that much for the preview: That's one big chunk of WASM:-)

**Reply by hunger** [reply]
@Frank1126lin: Your `Photo` is an a layout, but your `Image` is not: It's in Photo... So it defaults to `fill`. Setting the `height` based on the `width` does not change that: It just makes `Image` higher in such a way the aspect ratio is preserved. Then the `Image` gets clipped back down again to fit into the area of your `Photo`. That's why top/bottom get cut off.

**Reply by ogoffart** [reply]
@bjorn  it seems like this is a browser-dependent bug in the image rendering...  
`image-fit: fill` doesn't seem to work on firefox.  But it works with chrome, and also on every backend on desktop :-/
